modules/consensus/synchronize.go

```diff
@@ -389,13 +403,18 @@ func (cs *ConsensusSet) rpcRelayHeader(conn modules.PeerConn) error {
 	} else if err != nil {
 		return err
 	}
-	// If the header is valid and extends the heaviest chain, fetch, accept it,
-	// and broadcast it.
+
+	// If the header is valid and extends the heaviest chain, fetch the
+	// corresponding block. Call needs to be made in a separate goroutine
+	// because an exported call to the gateway is used, which is a deadlock
+	// risk given that rpcRelayHeader is called from the gateway.
+	wg.Add(1)
 	go func() {
-		err := cs.gateway.RPC(conn.RPCAddr(), "SendBlk", cs.threadedReceiveBlock(h.ID()))
+		err = cs.gateway.RPC(conn.RPCAddr(), "SendBlk", cs.threadedReceiveBlock(h.ID()))
 		if err != nil {
 			cs.log.Debugln("WARN: failed to get header's corresponding block:", err)
 		}
+		wg.Done()
```

Comment:
this is bizarre. You don't need a WaitGroup. Just move the `tg.Done` inside this function. It's fine as long as it's properly documented.